[DPE-4897]: Stabilize sharding race condition tests

Gu1nness opened the pull request:
## Issue

 * The sharding race condition test is flaky.
 * The test and the code seemed good on the Sharding side.
 * There was actually another race condition which was making this test flaky.

## Solution

 * The race condition that was happening is the following: the config server can run the `RelationJoined` hook before the shard has run `RelationCreated`. The `RelationJoined` hook relies on some information sent by the `RelationCreated` hook to succeed. The solution is hence to defer this event in the case we have some incompatible versions between shards and config server, or if we are missing those shards. This is done also for the shard because otherwise, the same issue could happen, we dropped the event and would never sync the credentials.
 * This patch reuses the already existing code as much at it can.
 * Adds some tests to ensure that this never happens againt

## Miscellaneous

 * Fix some small issues in the `test_config_server_lib.py` tests.

Gu1nness committed:
feat: Bump data_interfaces library to last version

Gu1nness committed:
fix: Lower update-status-hook-interval to 60s to avoid timeouts

Gu1nness committed:
style: use walrus operator + type hints

Gu1nness committed:
test: Bump delay to 2m instead of 60s

Gu1nness committed:
test: Setup upterm server

Gu1nness committed:
test: add some debug-log

Gu1nness committed:
fix: log & higher delay?

Gu1nness committed:
test: Add more debug log to track

Gu1nness committed:
fix: Always defer on mismatched versions

In case we have mismatched version/missing version during the integration of a cluster,
we always want to defer the `RelationJoined` event:
 * We let the operator upgrade the components
 * Integration resumes as expected after the upgrade has completed.

Dropping the event as we used to do was leading the integration to never
being fully completed in those cases:
 * The shard hasn't run it's `RelationCreated` event yet so we are missing
   the version information (which dropped the event, so never
   integrated, as we would never send credentials, hence a deadlock).
 * There's a cluster mismatch, hence we drop the event (same) and the
   credentials are never sent (hence another deadlock after the ugprade).

Both those cases would force the operator to remove the integration and
re-integrate, which is not what we want.

Gu1nness committed:
fix: typo

Gu1nness committed:
fix: Use status name

Gu1nness committed:
fix: other typo

Gu1nness committed:
test: Add test confirming the deferral

Gu1nness committed:
test: Final fix + style

Gu1nness committed:
test: Add a test against shards

Gu1nness committed:
fix: Useless here

Gu1nness committed:
Revert "test: Setup upterm server"

This reverts commit 4befe091be2b0a6a38e540fd3dc00f0681b413d1.

Gu1nness committed:
Revert "test: add some debug-log"

This reverts commit cc93970b7ceec03defde5293a7d88eb0222c5a95.

Gu1nness committed:
style: Remove debug log

MiaAltieri commented on lib/charms/mongodb/v1/shards_interface.py:
```diff
@@ -153,7 +153,13 @@ def pass_sanity_hook_checks(self, event: EventBase) -> bool:
 
         if not self.charm.is_relation_feasible(self.relation_name):
-            logger.info("Skipping event %s , relation not feasible.", str(type(event)))
+            if self.charm.status.is_status_related_to_mismatched_revision(
+                self.charm.unit.status.name
+            ):
+                logger.info("Deferring %s. Mismatched versions in the cluster.", str(type(event)))
+                event.defer()
+            else:
+                logger.info("Skipping event %s , relation not feasible.", str(type(event)))
```
Nit: Removing nested structure of if statement
```suggestion
      if self.charm.status.is_status_related_to_mismatched_revision(
           self.charm.unit.status.name
      ):
            logger.info("Deferring %s. Mismatched versions in the cluster.", str(type(event)))
            event.defer()
       elif not self.charm.is_relation_feasible(self.relation_name):
           logger.info("Skipping event %s , relation not feasible.", str(type(event)))
```

Gu1nness replied:
The suggestion doesn't work because the call to `self.charm.status.is_status_related_to_mismatched_revision` requires `self.charm.is_relation_feasible` to [have run](https://github.com/canonical/mongodb-operator/blob/6/edge/src/charm.py#L1621-L1623).
However I agree that this can be improved and I will come up with a not-nested version of this.

MiaAltieri commented on lib/charms/mongodb/v1/shards_interface.py:
```diff
@@ -764,7 +770,13 @@ def pass_sanity_hook_checks(self, event: EventBase) -> bool:
 
         if not self.charm.is_relation_feasible(self.relation_name):
-            logger.info("Skipping event %s , relation not feasible.", str(type(event)))
+            if self.charm.status.is_status_related_to_mismatched_revision(
+                self.charm.unit.status.name
+            ):
+                logger.info("Deferring %s. Mismatched versions in the cluster.", str(type(event)))
+                event.defer()
+            else:
+                logger.info("Skipping event %s , relation not feasible.", str(type(event)))
```
Nit: Removing nested structure of if statement
```suggestion
      if self.charm.status.is_status_related_to_mismatched_revision(
           self.charm.unit.status.name
      ):
            logger.info("Deferring %s. Mismatched versions in the cluster.", str(type(event)))
            event.defer()
       elif not self.charm.is_relation_feasible(self.relation_name):
           logger.info("Skipping event %s , relation not feasible.", str(type(event)))
```

Gu1nness replied:
(Same comment as the former one)

MiaAltieri commented on src/charm.py:
```diff
@@ -1075,7 +1076,7 @@ def install_snap_packages(self, packages: List[Package]) -> None:
     def _instatiate_keyfile(self, event: StartEvent) -> None:
         # wait for keyFile to be created by leader unit
-        if not self.get_secret(APP_SCOPE, Config.Secrets.SECRET_KEYFILE_NAME):
+        if not (keyfile := self.get_secret(APP_SCOPE, Config.Secrets.SECRET_KEYFILE_NAME)):
```
nice 😎

Gu1nness committed:
feat: Bump charm internal version

MiaAltieri commented on src/charm.py:
```diff
@@ -1661,6 +1662,8 @@ def get_cluster_mismatched_revision_status(self) -> Optional[StatusBase]:
                 return BlockedStatus(
                     f"Charm revision ({current_charms_version}{local_identifier}) is not up-to date with config-server."
                 )
+            if self.is_role(Config.Role.CONFIG_SERVER):
+                return WaitingStatus("No information received yet from at least one shard.")
```
So this is a tricky one, we could have a `NoVersionError` due to:
1. race-condition, i.e. waiting for the information to be shared through `RelationCreatedEvent`
2. the shards are an older revision that do not share their status

In case 1. the situation should resolve itself fairly quickly, if there is a race condition causing this, it is because the users immediately related the applications and they are waiting to "settle" in this case we will receive a flood of juju events and it should resolve itself quite quickly. 

In case 2. the situation will not resolve itself without the user doing something on the shard side.

In general we want to avoid setting "unnecessary statuses", determining which statuses are unnecessary is an art and I am still by no means perfect at this. 

For these reasons I would suggest to not report a status for case 1. and for case 2. report the following.

```suggestion
            if self.is_role(Config.Role.CONFIG_SERVER):
                return WaitingStatus(
                f"Waiting for shards to upgrade/downgrade to revision {current_charms_version}{local_identifier}."
            )
```

If you **really wanted to** you could add some sort of sanity check 
```
            if self.is_role(Config.Role.CONFIG_SERVER) and <has shard already started>:
                return WaitingStatus( f"Waiting for shards to upgrade/downgrade to revision {current_charms_version}{local_identifier}."
```
but this might be a little overkill

Gu1nness replied:
We can keep the code as it was then because the status will be handled by the code below this `try/catch`.
I'll remove this part!

MiaAltieri reviewed (changes requested):
Looks great! Thank you so much for your work here! Super appreciative!

Gu1nness committed:
fix: Remove code nesting

Also remove a useless and misleading status (see review comment).
This adds some code duplication between ShardingProvider and
ConfigServerRequirer. Maybe it could be good to see the level of code
duplication in those two classes, it could be abstracted to reuse the
shared code.

MiaAltieri commented on lib/charms/mongodb/v1/shards_interface.py:
```diff
@@ -764,8 +778,7 @@ def pass_sanity_hook_checks(self, event: EventBase) -> bool:
         if not self.charm.is_relation_feasible(self.relation_name):
-            logger.info("Skipping event %s , relation not feasible.", str(type(event)))
-            return False
+            self.__handle_relation_not_feasible(event)
```
move this check to after `if not self.charm.is_role(Config.Role.SHARD):`

reason: `__handle_relation_not_feasible ` defers, we don't want to end up in an infinite defer cycle if the component is not a shard. (same for above)

Gu1nness replied:
Done!

Gu1nness replied:
Actually, this has to be higher in the code as there are some checks that won't be run and some statuses won't be set (including the ones on the relation status and the current role) so I will be reverting this last commit and we will not end up in a infinite defer loop thanks to the checks in `is_relation_feasible`

Gu1nness committed:
fix: Move function below role check

MiaAltieri reviewed (dismissed):
Great work! thank you very much Neha!

Gu1nness committed:
Revert "fix: Move function below role check"

This reverts commit ba5de6b313d8325c3b62d7edcdf4cae4a4f50381.

Reason: `is_relation_feasible` runs many checks including some that set
statuses that would be discarded otherwise. We already have some tests
ensuring that those statuses are set and they also prevent to fall into
an infinite defer loop that was feared.

Mehdi-Bendriss commented on lib/charms/mongodb/v1/shards_interface.py:
```diff
@@ -145,6 +145,13 @@ def _on_relation_joined(self, event):
 
         self.database_provides.update_relation_data(event.relation.id, relation_data)
 
+    def __handle_relation_not_feasible(self, event: EventBase):
```
**nit:** though the `__` is commonly used for name mangling of "private" methods, it is not really necessary in the context of charms / charm libs, where inheritance is performed rather modestly, and where most of the client code is actual charm components aware of what (not) to use

Gu1nness replied:
Sure, I'll rename it then :)

Mehdi-Bendriss commented on lib/charms/mongodb/v1/shards_interface.py:
```diff
@@ -145,6 +145,13 @@ def _on_relation_joined(self, event):
         self.database_provides.update_relation_data(event.relation.id, relation_data)
 
+    def __handle_relation_not_feasible(self, event: EventBase):
+        if self.charm.status.is_status_related_to_mismatched_revision(self.charm.unit.status.name):
```
Given this method only performs a check/deferral - would it make sense to directly put the logic in `pass_sanity_hook_checks`? Or are you expecting this method to perform additional logic in the future?

Gu1nness replied:
I'm not expecting it to perform additional logic, I moved it from `pass_sanity_hook_checks` mainly to remove some nesting in the main function after [this](https://github.com/canonical/mongodb-operator/pull/435#discussion_r1687577879) comment from Mia.
I'm fine with both, I just find this code easier to read.

In the future, we should improve this `pass_sanity_hook_checks` and the way the `is_relation_feasible` works because `is_relation_feasible` (which looks like a `GET` function) does sets the status and it would be better that this function either returns nothing (AKA all is good) or a status that we can then handle and set. This helps decorrelating the functionality and makes a code easier to read and understand.

Mehdi-Bendriss reviewed (commented):
Thanks Neha! Just a minor question

Gu1nness committed:
style: Don't use private methods
